go/vt/sqlparser/normalizer.go

```diff
@@ -28,78 +28,150 @@ import (
-			// Make the bindvar
-			bval := sqlToBindvar(node)
-			if bval == nil {
-				// If unsuccessful continue.
-				return true, nil
-			}
-			// Check if there's a bindvar for that value already.
-			var key string
-			if bval.Type == sqltypes.VarBinary {
-				// Prefixing strings with "'" ensures that a string
-				// and number that have the same representation don't
-				// collide.
-				key = "'" + string(node.Val)
-			} else {
-				key = string(node.Val)
-			}
-			bvname, ok := vals[key]
-			if !ok {
-				// If there's no such bindvar, make a new one.
-				bvname, counter = newName(prefix, counter, reserved)
-				vals[key] = bvname
-				bindVars[bvname] = bval
-			}
-			// Modify the AST node to a bindvar.
-			node.Type = ValArg
-			node.Val = append([]byte(":"), bvname...)
-		case *ComparisonExpr:
-			switch node.Operator {
-			case InStr, NotInStr:
-			default:
-				return true, nil
-			}
-			// It's either IN or NOT IN.
-			tupleVals, ok := node.Right.(ValTuple)
-			if !ok {
-				return true, nil
-			}
-			// The RHS is a tuple of values.
-			// Make a list bindvar.
-			bvals := &querypb.BindVariable{
-				Type: querypb.Type_TUPLE,
-			}
-			for _, val := range tupleVals {
-				bval := sqlToBindvar(val)
-				if bval == nil {
-					return true, nil
-				}
-				bvals.Values = append(bvals.Values, &querypb.Value{
-					Type:  bval.Type,
-					Value: bval.Value,
-				})
-			}
-			var bvname string
-			bvname, counter = newName(prefix, counter, reserved)
-			bindVars[bvname] = bvals
-			// Modify RHS to be a list bindvar.
-			node.Right = ListArg(append([]byte("::"), bvname...))
+	nz := newNormalizer(stmt, bindVars, prefix)
+	_ = Walk(nz.WalkStatement, stmt)
+}
+
+type normalizer struct {
+	stmt     Statement
+	bindVars map[string]*querypb.BindVariable
+	prefix   string
+	reserved map[string]struct{}
+	counter  int
+	vals     map[string]string
+}
+
+func newNormalizer(stmt Statement, bindVars map[string]*querypb.BindVariable, prefix string) *normalizer {
+	return &normalizer{
+		stmt:     stmt,
+		bindVars: bindVars,
+		prefix:   prefix,
+		reserved: GetBindvars(stmt),
+		counter:  1,
+		vals:     make(map[string]string),
+	}
+}
+
+// WalkStatement is the top level walk function.
+// If it encounters a Select, it switches to a mode
+// where variables are deduped.
+func (nz *normalizer) WalkStatement(node SQLNode) (bool, error) {
+	switch node := node.(type) {
+	case *Select:
+		_ = Walk(nz.WalkSelect, node)
+		// Don't continue
+		return false, nil
+	case *SQLVal:
+		nz.convertSQLVal(node)
+	case *ComparisonExpr:
+		nz.convertComparison(node)
+	}
+	return true, nil
+}
+
+// WalkSelect normalizes the AST in Select mode.
+func (nz *normalizer) WalkSelect(node SQLNode) (bool, error) {
+	switch node := node.(type) {
+	case *SQLVal:
+		nz.convertSQLValDedup(node)
+	case *ComparisonExpr:
+		nz.convertComparison(node)
+	}
+	return true, nil
+}
+
+func (nz *normalizer) convertSQLValDedup(node *SQLVal) {
+	// If value is too long, don't dedup.
```

Comment:
Is the reason for this because the comparison is likely to be expensive CPU wise? If so we should add an additional comment indicating why.

Reply:
Added.